bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -2,1383 +2,708 @@
+        loop {
+            {
+                let mut guard = self.inner.lock().unwrap();
+                if let Some(guest_event) = guard.guest_events.pop_front() {
+                    return InputQueueEvent::GuestEvent(guest_event);
+                } else if let Some(req) = guard.external_requests.pop_front() {
+                    return InputQueueEvent::ExternalRequest(req);
+                }
+            }
 
-    /// Sets the published instance and/or migration state and increases the
-    /// state generation number.
-    fn set_published_state(&mut self, state: PublishedState) {
-        let (instance_state, migration_state) = match state {
-            PublishedState::Instance(i) => (Some(i), None),
-            PublishedState::Migration(m) => (None, Some(m)),
-            PublishedState::Complete(i, m) => (Some(i), Some(m)),
-        };
+            // Notifiers in this module must use `notify_one` so that their
+            // notifications will not be lost if they arrive after this routine
+            // checks the queues but before it actually polls the notify.
+            self.notify.notified().await;
+        }
+    }
 
-        let ApiMonitoredState {
-            state: old_state,
-            migration: old_migration,
-            ..
-        } = self.api_state_tx.borrow().clone();
+    /// Notifies the external request queue that the instance's state has
+    /// changed so that it can change the dispositions for new state change
+    /// requests.
+    fn notify_instance_state_change(
+        &self,
+        state: request_queue::InstanceStateChange,
+    ) {
+        let mut guard = self.inner.lock().unwrap();
+        guard.external_requests.notify_instance_state_change(state);
+    }
 
-        let state = instance_state.unwrap_or(old_state);
-        let migration = if let Some(migration_state) = migration_state {
-            migration_state.apply_to(old_migration)
-        } else {
-            old_migration
-        };
+    /// Submits an external state change request to the queue.
+    pub(super) fn queue_external_request(
+        &self,
+        request: ExternalRequest,
+    ) -> Result<(), request_queue::RequestDeniedReason> {
+        let mut inner = self.inner.lock().unwrap();
+        let result = inner.external_requests.try_queue(request);
+        if result.is_ok() {
+            self.notify.notify_one();
+        }
+        result
+    }
+}
 
-        info!(self.log, "publishing new instance state";
-              "gen" => self.state_gen,
-              "state" => ?state,
-              "migration" => ?migration);
+impl guest_event::VcpuEventHandler for InputQueue {
+    fn suspend_halt_event(&self, when: Duration) {
+        let mut guard = self.inner.lock().unwrap();
+        if guard
+            .guest_events
+            .enqueue(guest_event::GuestEvent::VcpuSuspendHalt(when))
+        {
+            self.notify.notify_one();
+        }
```

Comment:
nit, take it or leave it: this gets repeated enough times that _maybe_ we should have a 
```rust
impl InputQueue {
    // ...
    fn enqueue_guest_event(&self, event: guest_event::GuestEvent) {
        if self.inner.lock().unwrap().guest_events.enqueue(event) {
             self.notify.notify_one()
        }
    }
    // ...
}
```

not a big deal either way, though.